### PR TITLE
docker-compose: use "docker compose" when not using older docker-compose command

### PR DESCRIPTION
**Pull request by oalbrigt**
Fixes https://github.com/ClusterLabs/resource-agents/issues/1974

### CODE REVIEW

File: heartbeat/docker-compose [hunk 1]
```diff
@@ -116,6 +116,8 @@ else
 	COMMAND="$OCF_RESKEY_binpath_default"
 fi
 
+[ "${COMMAND##*/}" = "docker" ] && COMMAND="$COMMAND compose"
```
Comment (minal3) [inline heartbeat/docker-compose:119]: Script defaults to _/usr/bin/docker-compose_.
$COMMAND is then set to _/usr/bin/docker-compose_
If I am correct, this line should be:

```suggestion
[ "${COMMAND##*/}" = "docker-compose" ] && COMMAND="${COMMAND%-compose} compose"
```

**Review by minal3 — CHANGES_REQUESTED** [review]

File: heartbeat/docker-compose (hunk 1 above)
Comment (oalbrigt) [reply heartbeat/docker-compose:119]: We also need to be backwards compatible, so can you run `docker --version` and `docker-compose --version` and post the output?

Then I'll be able to use that command to check if we need to add " compose" or not. It would also be helpful if you can run it with v1 of docker compose as well if you have access to nodes with it.

Comment (minal3) [reply heartbeat/docker-compose:119]: I see, backwards compatibility, you are right. Latest versions are;
```
murat@phobos:~$ docker -v ; docker compose version
Docker version 24.0.7, build 24.0.7-0ubuntu4.1
Docker Compose version 2.24.6+ds1-0ubuntu2
```
I will go check my resources if I can install docker compose V1.

Comment (minal3) [reply heartbeat/docker-compose:119]: Docker Compose V1:
`docker-compose version 1.29.2, build unknown`

Comment (minal3) [reply heartbeat/docker-compose:119]: A suggestion: Separate RAs for V1 & V2? Will simplify the script code...

Comment (oalbrigt) [reply heartbeat/docker-compose:119]: That would just be more code to maintain.

Is there one parameter that works for all the versions? Like docker --version, docker compose --version, docker-compose --version?

Comment (minal3) [reply heartbeat/docker-compose:119]: They have done a serious upgrade on docker compose. This is from official documentation: _Unlike Compose V1, Compose V2 integrates into the Docker CLI platform and the recommended command-line syntax is docker compose._

I rechecked; _version_ syntax is strictly different. I could not find a common syntax in between.

How about introducing a new parameter to RA to specify version?

**Comment by oalbrigt** [comment]
retest this please

**Comment by oalbrigt** [comment]
> I saw your commit; I guess this issue is resolved?
The updated patch should fix it.

It adds " compose" to the command when "docker-compose version 1\."  isnt output when running the command with -v.

**Comment by minal3** [comment]
Yes I have seen the conditional statement: I will run the patched RA at my test cluster and test.

**Comment by oalbrigt** [comment]
retest this please

**Comment by minal3** [comment]
Unfortunately this fix fails. I created two resources on a test cluster:
```
murat@node1:~$ sudo pcs resource
  * res10_vip   (ocf:heartbeat:IPaddr2):         Started node1.dc.lab
  * res20_httpd (ocf:heartbeat:docker-compose):  Stopped
```
ocf:heartbeat:docker-compose fails as;

```
Failed Resource Actions:
  * res20_httpd start on node1.dc.lab returned 'not installed' (Setup problem: couldn't find command: /usr/bin/docker-compose) at Mon Sep 16 20:56:34 2024 after 30ms
```

I extracted ONLY the executing lines until right after the fix:
```
#!/bin/sh
##########################################################################
# Initialization:

: ${OCF_ROOT:=/usr/lib/ocf}
: ${OCF_FUNCTIONS_DIR=${OCF_ROOT}/lib/heartbeat}
#. ${OCF_FUNCTIONS_DIR}/ocf-shellfuncs

# Defaults
OCF_RESKEY_binpath_default=/usr/bin/docker-compose
OCF_RESKEY_ymlfile_default=docker-compose.yml
: ${OCF_RESKEY_binpath=${OCF_RESKEY_binpath_default}}
: ${OCF_RESKEY_ymlfile=${OCF_RESKEY_ymlfile_default}}

if [ -r "$OCF_RESKEY_binpath" -a -x "$OCF_RESKEY_binpath" ]; then
	COMMAND="$OCF_RESKEY_binpath"
else
	COMMAND="$OCF_RESKEY_binpath_default"
fi

#$COMMAND -v | grep -q "^docker-compose version 1\." || COMMAND="$COMMAND compose"
$COMMAND -v | grep -q "^docker-compose version 1\." || COMMAND="${COMMAND%%-compose} compose"
echo "$COMMAND"
```
Penultimate line is my modification. Just execute them, simple enough. HOWEVER, resource fails with my modification, issuing the same error;

```
Failed Resource Actions:
  * res20_httpd start on node1.dc.lab returned 'not installed' (Setup problem: couldn't find command: /usr/bin/docker-compose) at Mon Sep 16 20:56:34 2024 after 30ms
```

This means RA execution already complains about the NON-EXISTENT docker-compose BEFORE the fix.
I ran ocf-tester, it got angry;

```
murat@node1:~$ sudo ocf-tester -n ocf:heartbeat:docker-compose /usr/lib/ocf/resource.d/heartbeat/docker-compose 
Beginning tests for /usr/lib/ocf/resource.d/heartbeat/docker-compose...
/usr/lib/ocf/resource.d/heartbeat/docker-compose: 119: /usr/bin/docker-compose: not found
/usr/lib/ocf/resource.d/heartbeat/docker-compose: 119: /usr/bin/docker-compose: not found
* rc=5: Validation failed.  Did you supply enough options with -o ?
/usr/lib/ocf/resource.d/heartbeat/docker-compose: 119: /usr/bin/docker-compose: not found
ocf-exit-reason:Setup problem: couldn't find command: /usr/bin/docker-compose
Aborting tests
```

**Comment by oalbrigt** [comment]
I have one more fix to push before you can do further testing.

**Comment by oalbrigt** [comment]
I think I managed to cover all the corner-cases now.

**Comment by oalbrigt** [comment]
I added an update to also avoid error being output if the command doesnt exist.

**Comment by minal3** [comment]
Hello,
Patch is working fine now.
First I ran it against ocf-tester;
```
root@node1: ~
> ocf-tester -n ocf:heartbeat:docker-compose -o dirpath=/opt/dockerfiles/revproxy-nginx -o ymlfile=docker-compose.yaml /usr/lib/ocf/resource.d/heartbeat/docker-compose 
Beginning tests for /usr/lib/ocf/resource.d/heartbeat/docker-compose...
* Your agent does not support the notify action (optional)
* Your agent does not support the demote action (optional)
* Your agent does not support the promote action (optional)
* Your agent does not support promotable clones (optional)
* Your agent does not support the reload action (optional)
/usr/lib/ocf/resource.d/heartbeat/docker-compose passed all tests
```
My test cluster includes 4 nodes, one being a qdevice. All non-witness nodes have an nginx docker container image & the same docker-compose.yaml file. All nodes have ocf:heartbeat:docker-compose patched with your latest fix.

Result: I moved 

> Resource: res20_httpd (class=ocf provider=heartbeat type=docker-compose)

among all three. RA is executing docker-compose as expected. No errors, nginx container is running at any node, I was able to access default web page over VIP address.

**Comment by minal3** [comment]
Two items to discuss;

1. I have NOT done a test with docker-compose v1.29, after all this thorough work, it should be done.

2. Default value of attribute "ymlfile" is docker-compose.yml, hence outdated. This is from official Compose documentation;

_The default path for a Compose file is compose.yaml (preferred) or compose.yml that is placed in the working directory. Compose also supports docker-compose.yaml and docker-compose.yml for backwards compatibility of earlier versions. If both files exist, Compose prefers the canonical compose.yaml._

This is the resource definition at my test cluster;

> Resource: res20_httpd (class=ocf provider=heartbeat type=docker-compose)
>   Attributes: res20_httpd-instance_attributes
>     dirpath=/opt/dockerfiles/revproxy-nginx/
>     ymlfile=docker-compose.yaml
>   Operations:
>     monitor: res20_httpd-monitor-interval-60s
>       interval=60s timeout=10s
>     start: res20_httpd-start-interval-0s
>       interval=0s timeout=240s
>     stop: res20_httpd-stop-interval-0s
>       interval=0s timeout=20s

I deleted the attribute `ymlfile` and resource failed, it should be defaulting to `.yml` extension. Do you think the default value should be updated to `.yaml`?

Or an existence test: `[ -a docker-compose.yaml ] || [ -a docker-compose.yml ]`
